refactor UI markup and add support for BIP21

talvasconcelos opened the pull request:
...

Related to lnbits/lnbits#1812

motorina0 commented on templates/satspay/display.html:
```diff
@@ -104,9 +175,9 @@
+      </div> -->
     </q-card>
-    <q-card class="q-pa-lg" v-if="lnbtc">
+    <!-- <q-card class="q-pa-lg" v-if="lnbtc">
```
remove dead code

motorina0 commented on templates/satspay/display.html:
```diff
@@ -249,6 +371,12 @@
         return hostname
+      },
+      unifiedQR() {
+        let bitcoin = this.charge.onchainaddress && this.charge.onchainaddress.toUpperCase()
```
```suggestion
const bitcoin = (this.charge.onchainaddress || '').toUpperCase()
```

motorina0 commented on templates/satspay/display.html:
```diff
@@ -249,6 +371,12 @@
+        let bitcoin = this.charge.onchainaddress && this.charge.onchainaddress.toUpperCase()
+        let lightning = this.charge.payment_request && this.charge.payment_request.toUpperCase()
+        let amount = this.charge.amount / 1e8
+        return `bitcoin:${bitcoin}?amount=${amount}&lightning=${lightning}`
```
- what does it show if `${bitcoin}` or `${lightning}` are undefined?
   - is that in accordance with the BIP21 spec?

talvasconcelos replied:
according to the spec (https://bitcoinqr.dev/) it should show just the bitcoin address... i'm still figuring it out... focusing on getting the markup nicer.

motorina0 commented on templates/satspay/display.html:
```diff
@@ -221,6 +292,56 @@
 <script>
   Vue.component(VueQrcode.name, VueQrcode)
 
+  Vue.component('satspay-paid',
```
why not have the components in their own files?

talvasconcelos replied:
I'm just now learning how to do components... may do that further ahead!

talvasconcelos committed:
refactor UI markup and add support for BIP21

Update templates/satspay/display.html

Co-authored-by: Vlad Stan <stan.v.vlad@gmail.com>
format and clean up

fix bug and move components to external js

refactor UI markup and add support for BIP21

Update templates/satspay/display.html

Co-authored-by: Vlad Stan <stan.v.vlad@gmail.com>

talvasconcelos commented:
Closes #25

talvasconcelos commented:
Card now looks like this, with Unified QR code if onchain is used:

![uqr](https://i.imgur.com/dDRk4GY.png)

Time elapsed:
![time](https://i.imgur.com/kg19mbK.png)

Paid:
![paid](https://i.imgur.com/B69ph36.png)

motorina0 commented:
- `mempool.js` was added in this PR https://github.com/lnbits/satspay/pull/26
- please remove it from this PR, it will make it easier to review

motorina0 commented on migrations.py:
```diff
@@ -93,3 +93,22 @@ async def m006_add_zeroconf_column(db):
         pass
+
+
+async def m007_add_pending_column(db):
```
- if I understand correctly this change is unrelated to the this PR
   - it was supposed to be here: https://github.com/lnbits/satspay/pull/22/files#diff-d7af58842a76f9c7d30d94d9e2539190b7f4d85a238ef85dc70e4dad3b026884
- we should have a release with this change alone ASAP (since it is a blocker bug)

motorina0 replied:
hmm, I can see this is on main already

motorina0 commented on migrations.py:
```diff
@@ -93,3 +93,22 @@ async def m006_add_zeroconf_column(db):
+            f"ALTER TABLE satspay.charges ADD COLUMN pending {db.big_int} NOT NULL DEFAULT 0;"
+        )
+
+        await db.execute(
```
- I think this is redundant,  the value should be `0` by default

talvasconcelos replied:
just the `ALTER TABLE...` then , right?

talvasconcelos committed:
format and clean up

fix bug and move components to external js

use qr code with logo

better markup

more markup tweaks

serve mempooljs from extension

html markup major refactor

Revert "serve mempooljs from extension"

This reverts commit d47888547c466971211bc17081c8ecb78589414f.

Revert "Merge remote-tracking branch 'origin/refactor_ui_UQR_BIP21' into refactor_ui_UQR_BIP21"

This reverts commit ae9cb4e0d267d1e185f7fbee738dc9ebbdf49832, reversing
changes made to d1ca31ae5af0263cdd3d93c2a495f7834c7f46c8.

remove unused code

talvasconcelos committed:
undo deleted file

motorina0 commented on static/js/components.js:
```diff
@@ -0,0 +1,96 @@
+    </div>
+    `,
+  created() {
+    console.log(this.charge.time_left)
```
- remove line

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,360 +1,234 @@
 {% endblock %} {% block scripts %}
 
-<script src="{{ url_for('satspay_static', path='js/mempool.js') }}"></script>
+<script src="https://mempool.space/mempool.js"></script>
```
- merge error
```suggestion
<script src="{{ url_for('satspay_static', path='js/mempool.js') }}"></script>
```

motorina0 commented on static/js/components.js:
```diff
@@ -0,0 +1,96 @@
+Vue.component('satspay-paid', {
+  props: ['charge', 'timer'],
+  template: `
+    <div class="col-sm-10 col-lg-8">
```
- the component should not force any col size
- having a `col` at top level also means the parent component has to have a `row`
```suggestion
    <div>
```

talvasconcelos replied:
i see your point. but these are specific components. removing the col size will make the QR smaller than "optimal"!

would it be ok if i pass the `class="col-sm-10 col-lg-8"` to the component? 

`<satspay-paid class="col-sm-10 col-lg-8">...</......>` ?

motorina0 replied:
why not just move the `<div col="...` at the parent level?

talvasconcelos replied:
i did

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,360 +1,234 @@
+                ></q-icon>
+              </div>
+              <div v-else-if="charge.paid">
+                <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
```
- move the column layout outside the `satspay-paid` component
```html
<div class="col-sm-10 col-lg-8"...
```

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,360 +1,234 @@
-                  >
-                </div>
+              <div v-else-if="charge.paid">
+                <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
```
- move the column layout outside the `satspay-paid` component
```html
<div class="col-sm-10 col-lg-8"...
```

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,360 +1,234 @@
-                  >
-                </div>
+              <div v-else-if="charge.paid">
+                <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
```
- move the column layout outside the `satspay-paid` component
```html
<div class="col-sm-10 col-lg-8"...
```

motorina0 commented on static/js/components.js:
```diff
@@ -0,0 +1,96 @@
+Vue.component('satspay-show-qr', {
+  props: ['charge', 'tab', 'value', 'href'],
+  template: `
+    <div class="col-sm-10 col-lg-8">
```
- see the comment on the root element of `'satspay-paid'`

motorina0 commented on static/js/components.js:
```diff
@@ -0,0 +1,96 @@
+    <div class="col-sm-10 col-lg-8">
+        <q-icon name="check" style="color: green; font-size: 21.4em" class="fit"></q-icon>
+        <div class="row text-center q-mt-lg">
+        <div class="col text-center">
```
- this `<div>` should be indented one level

motorina0 commented on static/js/components.js:
```diff
@@ -0,0 +1,96 @@
+        <q-icon name="check" style="color: green; font-size: 21.4em" class="fit"></q-icon>
+        <div class="row text-center q-mt-lg">
+        <div class="col text-center">
+            <q-btn outline v-if="charge.completelink" :loading="charge.paid" :percentage="timer" type="a"
```
- previously it was:
```html
<q-btn outline v-if="charge.webhook" 
```
- is the change on purpose (looks right)?

talvasconcelos replied:
yes it was wrong

motorina0 commented on static/js/components.js:
```diff
@@ -0,0 +1,96 @@
+})
+
+Vue.component('satspay-show-qr', {
+  props: ['charge', 'tab', 'value', 'href'],
```
- no need to pass the full charge data, the amount is sufficient
```suggestion
  props: ['charge-amount', 'tab', 'value', 'href'],
```

motorina0 replied:
- this component knows nothing about `tabs`, maybe the input should be named `type`
- `tab` is just an outer wrapper, in different contexts it can be something else (dialog, expandable-item, etc)

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,360 +1,234 @@
-              >
-                <q-tooltip> pay onchain </q-tooltip>
-              </q-btn>
+                :charge="charge"
```
```suggestion
                :charge-amount="charge.amount"
```

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,360 +1,234 @@
-                class="text-secondary"
+              <satspay-show-qr
+                v-else
+                :charge="charge"
```
```suggestion
                :charge-amount="charge.amount"
```

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,360 +1,234 @@
               </div>
+              <satspay-show-qr
+                v-else
+                :charge="charge"
```
```suggestion
                :charge-amount="charge.amount"
```

talvasconcelos committed:
Update templates/satspay/display.html

Co-authored-by: Vlad Stan <stan.v.vlad@gmail.com>

talvasconcelos committed:
resolve review issues

motorina0 commented on static/js/components.js:
```diff
@@ -0,0 +1,98 @@
+  created() {
+    if (!this.charge.timeLeft && !this.charge.paid) {
+      this.barText = 'Time elapsed'
+      return
```
```suggestion
```
- do not use return, later some more logic might be added to the end of `created()`

motorina0 commented on static/js/components.js:
```diff
@@ -0,0 +1,98 @@
+      return
+    } else if (this.charge.paid) {
+      this.barText = 'Payment received'
+      return
```
```suggestion
```
- do not use return, later some more logic might be added to the end of `created()`

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,329 +1,210 @@
+            <q-item-section side>
               <q-badge color="gray">
-                <span v-text="pendingFunds" class="text-subtitle2"></span> sat
+                <span class="text-subtitle2">${pendingFunds} sats</span>
```
- is the `${pendingFunds}` expression evaluated inside an HTML?

talvasconcelos replied:
It's javascript! It uses `delimiters` so we don't need to use `{% raw %}`

motorina0 replied:
Its JS inside HTML. I didn't know this works.

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,329 +1,210 @@
+          <q-item dense>
+            <q-item-section>
+              <q-item-label>Amount due:</q-item-label>
+              <!-- <q-item-label caption>Amount left to be paid</q-item-label> -->
```
- dead code
```suggestion
```

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,329 +1,210 @@
+            <q-item-section side>
+              <q-badge color="green">
+                <span class="text-subtitle2"
+                  >${charge.amount - charge.balance > 0 ? charge.amount -
```
- is `${...}` actually ok?

talvasconcelos replied:
it uses `delimiter` from vue, so we don't need the `{% raw %}` and can use JS

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,329 +1,210 @@
+          <q-item dense>
+            <q-item-section>
+              <q-item-label>Total to pay:</q-item-label>
+              <!-- <q-item-label caption>Total amount to pay</q-item-label> -->
```
```suggestion
```

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,329 +1,210 @@
+          <q-item dense>
+            <q-item-section>
+              <q-item-label>Amount paid:</q-item-label>
+              <!-- <q-item-label caption>Amount already paid</q-item-label> -->
```
```suggestion
```

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,329 +1,210 @@
+          <q-item v-if="pendingFunds" dense>
+            <q-item-section>
+              <q-item-label>Amount pending:</q-item-label>
+              <!-- <q-item-label caption>Amount pending on mempool</q-item-label> -->
```
```suggestion
```

motorina0 commented on templates/satspay/display.html:
```diff
@@ -367,6 +247,17 @@
+      unifiedQR() {
+        const bitcoin = (this.charge.onchainaddress || '').toUpperCase()
+        let queryString = `bitcoin:${bitcoin}?amount=${
+          this.charge.amount / 1e8
```
- JS floating point `toString()` can be tricky:
![image](https://github.com/lnbits/satspay/assets/2951406/7450d42f-859e-4c16-a189-7b3837fa7272)

talvasconcelos replied:
better to explicitly do `this.charge.amount / 100000000` ?

motorina0 replied:
- it would give the same exponential result
- suggestion: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed

motorina0 replied:
This is not resolved, please have a look t [toFixed]( https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed) function

talvasconcelos replied:
ahh, i see the difference! pushed, `( this.charge.amount / 1e8).toFixed(8)`

motorina0 commented on templates/satspay/display.html:
```diff
@@ -1,329 +1,210 @@
+          <q-tab-panel name="uqr">
+            <div class="row justify-center q-mt-sm">
+              <div class="col-sm-10 col-md-8">
+                <div v-if="!charge.timeLeft && !charge.paid">
```
- this conditions repeats 3 times (once in each tab)
- it should be moved on the `<q-tab` level

```html
<div v-if="!charge.timeLeft && !charge.paid">
    <q-icon
      name="block"
      style="color: #ccc; font-size: 21.4em"
    ></q-icon>
</div>
<div v-else-if="charge.paid">
    <satspay-paid :charge="charge" :timer="timer"></satspay-paid>
</div>
<q-tabs v-else ...>
   ...
</q-tabs>
```

motorina0 commented:
Scanning with Muun and Phoenix did not work. Maybe they don't support BIP21?

talvasconcelos commented:
> Scanning with Muun and Phoenix did not work. Maybe they don't support BIP21?

on Muun: "Go to Settings > Lightning Network > Receiving protocol and set it to"Unified" to enable this feature
Link to issue:"

Phoenix should work, according to this: https://bitcoinqr.dev/

I tried Phoenix, but don't have it on testnet! But it reads the QR

talvasconcelos committed:
Update static/js/components.js

Co-authored-by: Vlad Stan <stan.v.vlad@gmail.com>

talvasconcelos committed:
Update static/js/components.js

Co-authored-by: Vlad Stan <stan.v.vlad@gmail.com>

talvasconcelos committed:
address Vlad's comments

talvasconcelos committed:
you gotta love JS

dni reviewed (approved):
untested
